rpc: avoid use-after-Stop via RPC endpoints

tbg opened the pull request:
This avoids a class of crashes on shutdown that occur when an RPC is
handled after the stopper has already stopped (and in the process has
stopped the Stores' pebble instances. These crashes are probably very
rare in production but hit with some regularity in unit tests, for an example see:
https://github.com/cockroachdb/cockroach/issues/68395#issuecomment-975545855

I was also able to hit this locally within a few minutes by running

`make stress PKG=./pkg/kv/kvserver/ TESTS=TestStrictGCEnforcement/protected_timestamps_are_respected`.

Fixes #73154.

Release note: None

tbg committed:
rpc: rename all ctx receivers to rpcCtx

`ctx` should be reserved for `context.Context`.

Release note: None

tbg committed:
rpc: avoid use-after-Stop via RPC endpoints

This avoids a class of crashes on shutdown that occur when an RPC is
handled after the stopper has already stopped (and in the process has
stopped the Stores' pebble instances. These crashes are probably very
rare in production but hit with some regularity in unit tests, for an example see:
https://github.com/cockroachdb/cockroach/issues/68395#issuecomment-975545855

I was also able to hit this locally within a few minutes by running

`make stress PKG=./pkg/kv/kvserver/ TESTS=TestStrictGCEnforcement/protected_timestamps_are_respected`.

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/73155)
<!-- Reviewable:end -->

erikgrinaker reviewed (approved):
Is this going to meaningfully affect performance? Looks like it just adds on a mutex read lock for each RPC, which I suppose should be cheap enough given that this generally hits the network.

tbg commented:
> Is this going to meaningfully affect performance? Looks like it just adds on a mutex read lock for each RPC, which I suppose should be cheap enough given that this generally hits the network.

I don't think it should but I can check. What do you think is the best end-to-end test here? kv95?


----

I tried adding a test for this, but unfortunately it's not easy, at least not as an end-to-end test, due to the need to race the request into gRPC at the right time. Here's my latest attempt which did not work, curiously this results in a `200 OK` even though this really should error out as I'm asking for a nonexistent method. But this http gateway isn't something we ought to be using anyway so posting this only for completeness.

```go
func TestServer_UseAfterStop(t *testing.T) {
	defer leaktest.AfterTest(t)()
	defer log.Scope(t).Close(t)
	ctx := context.Background()

	s, err := serverutils.StartServerRaw(base.TestServerArgs{})
	require.NoError(t, err)

	s.Stopper().Stop(ctx)

	req, err := http.NewRequest(http.MethodGet, "http://127.0.0.1/AsdPing", strings.NewReader(""))
	req.ProtoMajor = 2
	req.Method = http.MethodPost
	req.Header.Add("Content-Type", "application/grpc")
	require.NoError(t, err)
	var resp httptest.ResponseRecorder
	s.(*TestServer).grpc.ServeHTTP(&resp, req)
	body, err := ioutil.ReadAll(resp.Result().Body)
	require.NoError(t, err)
	t.Errorf("%s %s", resp.Result().Status, body)
}

```

erikgrinaker commented:
>I don't think it should but I can check. What do you think is the best end-to-end test here? kv95?

Yeah, that should do it.

erikgrinaker commented:
>I tried adding a test for this, but unfortunately it's not easy, at least not as an end-to-end test, due to the need to race the request into gRPC at the right time.

I think collateral damage in other tests is probably signal enough here.

tbg commented:
Ugh, unfortunately this does look a lot worse. Might be noisy neighbors but definitely too suspicious to just ignore. I'll see if I can find something similar using a `SELECT single-row` type Go benchmark.

this PR (2f11e06d30edb67fbd1d2fe3451507022bcd9506):

```
==> artifacts/2f11e06d30/144045/kv95/enc=false/nodes=3/run_1/run_135736.860684236_n4_workload_run_kv.log <==
_elapsed___errors_____ops(total)___ops/sec(cum)__avg(ms)__p50(ms)__p95(ms)__p99(ms)_pMax(ms)__total
  600.0s        0       20711380        34518.8      5.0      3.7     15.7     26.2     92.3  read

_elapsed___errors_____ops(total)___ops/sec(cum)__avg(ms)__p50(ms)__p95(ms)__p99(ms)_pMax(ms)__total
  600.0s        0        1090075         1816.8     10.8     10.0     25.2     35.7    121.6  write

_elapsed___errors_____ops(total)___ops/sec(cum)__avg(ms)__p50(ms)__p95(ms)__p99(ms)_pMax(ms)__result
  600.0s        0       21801455        36335.6      5.3      3.8     16.8     27.3    121.6  
run_135736.860684236_n4_workload_run_kv: 14:07:49 cluster.go:2030: > result: <nil>

==> artifacts/2f11e06d30/144045/kv95/enc=false/nodes=3/run_2/run_135844.661843927_n4_workload_run_kv.log <==
_elapsed___errors_____ops(total)___ops/sec(cum)__avg(ms)__p50(ms)__p95(ms)__p99(ms)_pMax(ms)__total
  600.0s        0       20806092        34676.7      4.9      3.7     15.2     26.2   1677.7  read

_elapsed___errors_____ops(total)___ops/sec(cum)__avg(ms)__p50(ms)__p95(ms)__p99(ms)_pMax(ms)__total
  600.0s        0        1095396         1825.7     11.8     10.0     25.2     39.8   1677.7  write

_elapsed___errors_____ops(total)___ops/sec(cum)__avg(ms)__p50(ms)__p95(ms)__p99(ms)_pMax(ms)__result
  600.0s        0       21901488        36502.4      5.3      3.9     16.3     27.3   1677.7  
run_135844.661843927_n4_workload_run_kv: 14:08:56 cluster.go:2030: > result: <nil>

==> artifacts/2f11e06d30/144045/kv95/enc=false/nodes=3/run_3/run_140109.146360739_n4_workload_run_kv.log <==
_elapsed___errors_____ops(total)___ops/sec(cum)__avg(ms)__p50(ms)__p95(ms)__p99(ms)_pMax(ms)__total
  600.0s        0       21283687        35472.7      4.8      3.7     15.2     25.2    352.3  read

_elapsed___errors_____ops(total)___ops/sec(cum)__avg(ms)__p50(ms)__p95(ms)__p99(ms)_pMax(ms)__total
  600.0s        0        1118631         1864.4     11.3      9.4     25.2     37.7    369.1  write

_elapsed___errors_____ops(total)___ops/sec(cum)__avg(ms)__p50(ms)__p95(ms)__p99(ms)_pMax(ms)__result
  600.0s        0       22402318        37337.1      5.1      3.8     15.7     26.2    369.1  
run_140109.146360739_n4_workload_run_kv: 14:11:21 cluster.go:2030: > result: <nil>

==> artifacts/2f11e06d30/144045/kv95/enc=false/nodes=3/run_4/run_140149.826176564_n4_workload_run_kv.log <==
_elapsed___errors_____ops(total)___ops/sec(cum)__avg(ms)__p50(ms)__p95(ms)__p99(ms)_pMax(ms)__total
  600.0s        0       21439588        35732.6      4.8      3.5     15.2     25.2    121.6  read

_elapsed___errors_____ops(total)___ops/sec(cum)__avg(ms)__p50(ms)__p95(ms)__p99(ms)_pMax(ms)__total
  600.0s        0        1128605         1881.0     10.4      9.4     24.1     33.6    113.2  write

_elapsed___errors_____ops(total)___ops/sec(cum)__avg(ms)__p50(ms)__p95(ms)__p99(ms)_pMax(ms)__result
  600.0s        0       22568193        37613.6      5.1      3.8     16.3     26.2    121.6  
run_140149.826176564_n4_workload_run_kv: 14:12:01 cluster.go:2030: > result: <nil>

==> artifacts/2f11e06d30/144045/kv95/enc=false/nodes=3/run_5/run_135750.104347005_n4_workload_run_kv.log <==
_elapsed___errors_____ops(total)___ops/sec(cum)__avg(ms)__p50(ms)__p95(ms)__p99(ms)_pMax(ms)__total
  600.0s        0       21037822        35062.9      4.9      3.7     15.7     26.2    838.9  read

_elapsed___errors_____ops(total)___ops/sec(cum)__avg(ms)__p50(ms)__p95(ms)__p99(ms)_pMax(ms)__total
  600.0s        0        1106895         1844.8     10.7      9.4     25.2     35.7    805.3  write

_elapsed___errors_____ops(total)___ops/sec(cum)__avg(ms)__p50(ms)__p95(ms)__p99(ms)_pMax(ms)__result
  600.0s        0       22144717        36907.8      5.2      3.8     16.3     26.2    838.9  
run_135750.104347005_n4_workload_run_kv: 14:08:02 cluster.go:2030: > result: <nil>

```

master base commit (38fd4064f4):

```
tail -n 9 artifacts/38fd4064f4/144307/kv95/enc\=false/nodes\=3/run_*/run_*kv.log 
==> artifacts/38fd4064f4/144307/kv95/enc=false/nodes=3/run_1/run_140155.423500120_n4_workload_run_kv.log <==
_elapsed___errors_____ops(total)___ops/sec(cum)__avg(ms)__p50(ms)__p95(ms)__p99(ms)_pMax(ms)__total
  600.0s        0       21390209        35650.3      4.8      3.5     14.7     25.2    369.1  read

_elapsed___errors_____ops(total)___ops/sec(cum)__avg(ms)__p50(ms)__p95(ms)__p99(ms)_pMax(ms)__total
  600.0s        0        1124719         1874.5     11.2      9.4     25.2     37.7    352.3  write

_elapsed___errors_____ops(total)___ops/sec(cum)__avg(ms)__p50(ms)__p95(ms)__p99(ms)_pMax(ms)__result
  600.0s        0       22514928        37524.8      5.1      3.8     15.7     26.2    369.1  
run_140155.423500120_n4_workload_run_kv: 14:12:07 cluster.go:2030: > result: <nil>

==> artifacts/38fd4064f4/144307/kv95/enc=false/nodes=3/run_2/run_140113.076486501_n4_workload_run_kv.log <==
_elapsed___errors_____ops(total)___ops/sec(cum)__avg(ms)__p50(ms)__p95(ms)__p99(ms)_pMax(ms)__total
  600.0s        0       21317121        35528.4      4.9      3.7     15.2     25.2     96.5  read

_elapsed___errors_____ops(total)___ops/sec(cum)__avg(ms)__p50(ms)__p95(ms)__p99(ms)_pMax(ms)__total
  600.0s        0        1123676         1872.8     10.4      9.4     24.1     35.7     88.1  write

_elapsed___errors_____ops(total)___ops/sec(cum)__avg(ms)__p50(ms)__p95(ms)__p99(ms)_pMax(ms)__result
  600.0s        0       22440797        37401.2      5.1      3.8     16.3     26.2     96.5  
run_140113.076486501_n4_workload_run_kv: 14:11:25 cluster.go:2030: > result: <nil>

==> artifacts/38fd4064f4/144307/kv95/enc=false/nodes=3/run_3/run_140218.000871884_n4_workload_run_kv.log <==
_elapsed___errors_____ops(total)___ops/sec(cum)__avg(ms)__p50(ms)__p95(ms)__p99(ms)_pMax(ms)__total
  600.0s        0       21466249        35777.0      4.8      3.7     15.2     25.2    113.2  read

_elapsed___errors_____ops(total)___ops/sec(cum)__avg(ms)__p50(ms)__p95(ms)__p99(ms)_pMax(ms)__total
  600.0s        0        1128788         1881.3     10.4      9.4     24.1     35.7     92.3  write

_elapsed___errors_____ops(total)___ops/sec(cum)__avg(ms)__p50(ms)__p95(ms)__p99(ms)_pMax(ms)__result
  600.0s        0       22595037        37658.3      5.1      3.8     15.7     26.2    113.2  
run_140218.000871884_n4_workload_run_kv: 14:12:30 cluster.go:2030: > result: <nil>

==> artifacts/38fd4064f4/144307/kv95/enc=false/nodes=3/run_4/run_140009.503132035_n4_workload_run_kv.log <==
_elapsed___errors_____ops(total)___ops/sec(cum)__avg(ms)__p50(ms)__p95(ms)__p99(ms)_pMax(ms)__total
  600.0s        0       21426098        35710.1      4.8      3.7     15.2     25.2    109.1  read

_elapsed___errors_____ops(total)___ops/sec(cum)__avg(ms)__p50(ms)__p95(ms)__p99(ms)_pMax(ms)__total
  600.0s        0        1128114         1880.2     10.5      9.4     24.1     33.6     88.1  write

_elapsed___errors_____ops(total)___ops/sec(cum)__avg(ms)__p50(ms)__p95(ms)__p99(ms)_pMax(ms)__result
  600.0s        0       22554212        37590.3      5.1      3.8     15.7     26.2    109.1  
run_140009.503132035_n4_workload_run_kv: 14:10:21 cluster.go:2030: > result: <nil>

==> artifacts/38fd4064f4/144307/kv95/enc=false/nodes=3/run_5/run_140212.131244018_n4_workload_run_kv.log <==
_elapsed___errors_____ops(total)___ops/sec(cum)__avg(ms)__p50(ms)__p95(ms)__p99(ms)_pMax(ms)__total
  600.0s        0       21124242        35207.0      4.9      3.7     15.2     25.2     92.3  read

_elapsed___errors_____ops(total)___ops/sec(cum)__avg(ms)__p50(ms)__p95(ms)__p99(ms)_pMax(ms)__total
  600.0s        0        1111166         1851.9     10.6      9.4     24.1     35.7     83.9  write

_elapsed___errors_____ops(total)___ops/sec(cum)__avg(ms)__p50(ms)__p95(ms)__p99(ms)_pMax(ms)__result
  600.0s        0       22235408        37058.9      5.2      3.8     16.3     26.2     92.3  
run_140212.131244018_n4_workload_run_kv: 14:12:24 cluster.go:2030: > result: <nil>
```

tbg commented:
Did some quick microbenchmarking on the stopper. The cost of spawning a task is basically `RLock` and `RUnlock` and the atomic increment. I made sure this doesn't degenerate as concurrency goes up, at least not quickly:

```
cpu: AMD Ryzen 7 3800X 8-Core Processor             
BenchmarkStopperPar
BenchmarkStopperPar/parallelism=1
BenchmarkStopperPar/parallelism=1-16         	28944364	        39.80 ns/op
BenchmarkStopperPar/parallelism=4
BenchmarkStopperPar/parallelism=4-16         	29925592	        39.93 ns/op
BenchmarkStopperPar/parallelism=8
BenchmarkStopperPar/parallelism=8-16         	29886718	        39.93 ns/op
BenchmarkStopperPar/parallelism=16
BenchmarkStopperPar/parallelism=16-16        	29817416	        39.88 ns/op
BenchmarkStopperPar/parallelism=32
BenchmarkStopperPar/parallelism=32-16        	29977888	        39.99 ns/op
BenchmarkStopperPar/parallelism=64
BenchmarkStopperPar/parallelism=64-16        	30038031	        39.96 ns/op
BenchmarkStopperPar/parallelism=128
BenchmarkStopperPar/parallelism=128-16       	29478643	        39.91 ns/op
BenchmarkStopperPar/parallelism=256
BenchmarkStopperPar/parallelism=256-16       	29075911	        40.03 ns/op
BenchmarkStopperPar/parallelism=512
BenchmarkStopperPar/parallelism=512-16       	27592840	        39.96 ns/op
BenchmarkStopperPar/parallelism=1024
BenchmarkStopperPar/parallelism=1024-16      	25764234	        40.07 ns/op
BenchmarkStopperPar/parallelism=2048
BenchmarkStopperPar/parallelism=2048-16      	29895319	        40.12 ns/op
BenchmarkStopperPar/parallelism=4096
BenchmarkStopperPar/parallelism=4096-16      	28818667	        40.22 ns/op
BenchmarkStopperPar/parallelism=8192
BenchmarkStopperPar/parallelism=8192-16      	28192084	        40.40 ns/op
BenchmarkStopperPar/parallelism=16384
BenchmarkStopperPar/parallelism=16384-16     	27473606	        42.09 ns/op
PASS
```

That last result uses 16384*16 goroutines, so way in excess of what we can expect to see in kv95. I might have to re-run the kv95 experiment as there's no way this can be explained. I did do an audit of the stopper to see if we perhaps might end up exclusively locking the mutex frequently (I was suspecting `WithCancelOnQuiesce` for example) but no dice. The task really shouldn't be adding anything like the overhead my experiment shows so best guess is that I got unlucky when testing this PR.

knz reviewed (commented):
LGTM on the approach.

Re the performance, could it be the overhead of the extra heap traffic due to the closure on every request?

> Reviewed 1 of 1 files at r1, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/73155#-:-MpNRq3_8QTbLrU2tQXF:b-iriavr)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @knz and @otan)



<!-- Sent from Reviewable.io -->

tbg commented:
> Re the performance, could it be the overhead of the extra heap traffic due to the closure on every request?

At least according to `./scripts/goescape ./pkg/rpc/context.go` they don't actually escape, so there should be zero difference in allocations.

erikgrinaker commented:
> I might have to re-run the kv95 experiment as there's no way this can be explained.

I think it's probably fine, I agree that this shouldn't really have a significant impact here. I'd go ahead and merge, and roachperf will tell if there's any effect.

tbg commented:
Since `roachtest` is effectively down and might not be fixed until next week, I'll hold off. I might re-trigger the comparison later and hope for a better outcome.

tbg commented:
Ran the kv95 tests again at got similar results.

master base:

```
_elapsed___errors_____ops(total)___ops/sec(cum)__avg(ms)__p50(ms)__p95(ms)__p99(ms)_pMax(ms)__result
  600.0s        0       22739156        37898.5      5.1      3.8     15.7     26.2   1677.7  
  600.0s        0       22798360        37997.2      5.1      3.8     15.7     26.2   1610.6  
  600.0s        0       22637805        37729.5      5.1      3.8     15.7     26.2    117.4  
  600.0s        0       22803928        38006.5      5.1      3.8     15.7     26.2    104.9  
  600.0s        0       22624234        37706.9      5.1      3.8     15.7     26.2    113.2  
```

PR:

```
_elapsed___errors_____ops(total)___ops/sec(cum)__avg(ms)__p50(ms)__p95(ms)__p99(ms)_pMax(ms)__result
  600.0s        0       22108088        36846.6      5.2      3.9     16.3     26.2    209.7  
  600.0s        0       22050947        36751.5      5.2      3.9     16.3     26.2    125.8  
  600.0s        0       22533830        37556.3      5.1      3.8     15.7     26.2    872.4  
  600.0s        0       22278315        37130.4      5.2      3.8     16.3     26.2    402.7  
  600.0s        0       22374288        37290.4      5.1      3.8     16.3     26.2    121.6  
```

Time to actually hit the multi-node microbenchmarks.

tbg commented:
old is master base, new is PR

```
$ benchstat 38fd4064f40c574d0b2aee02f414787a2760fcec.txt 2f11e06d30edb67fbd1d2fe3451507022bcd9506.txt 
name                                        old time/op    new time/op    delta
Scan/MultinodeCockroach/count=1/limit=1-16     488µs ± 2%     493µs ± 2%   ~     (p=0.310 n=5+5)

name                                        old alloc/op   new alloc/op   delta
Scan/MultinodeCockroach/count=1/limit=1-16    64.5kB ± 1%    64.2kB ± 0%   ~     (p=0.310 n=5+5)

name                                        old allocs/op  new allocs/op  delta
Scan/MultinodeCockroach/count=1/limit=1-16       751 ± 0%       751 ± 0%   ~     (all equal)
```

getting about the same when I disable the local server optimization,

```diff
diff --git a/pkg/rpc/context.go b/pkg/rpc/context.go
index dc5d62f998..5eeebf2fbb 100644
--- a/pkg/rpc/context.go
+++ b/pkg/rpc/context.go
@@ -696,7 +696,7 @@ func IsLocal(iface roachpb.InternalClient) bool {
 
 // SetLocalInternalServer sets the context's local internal batch server.
 func (ctx *Context) SetLocalInternalServer(internalServer roachpb.InternalServer) {
-	ctx.localInternalClient = internalClientAdapter{internalServer}
+	//ctx.localInternalClient = internalClientAdapter{internalServer}
 }
 
 // removeConn removes the given connection from the pool. The supplied connKeys

```

```
tobias@td:~/go/src/github.com/cockroachdb/cockroach$ benchstat 38fd4064f40c574d0b2aee02f414787a2760fcec.txt 2f11e06d30edb67fbd1d2fe3451507022bcd9506.txt 
name                                        old time/op    new time/op    delta
Scan/MultinodeCockroach/count=1/limit=1-16     664µs ± 1%     668µs ± 0%   ~     (p=0.056 n=5+5)

name                                        old alloc/op   new alloc/op   delta
Scan/MultinodeCockroach/count=1/limit=1-16    83.0kB ± 1%    82.9kB ± 1%   ~     (p=0.841 n=5+5)

name                                        old allocs/op  new allocs/op  delta
Scan/MultinodeCockroach/count=1/limit=1-16       993 ± 0%       993 ± 0%   ~     (p=1.000 n=4+4)
```

tbg commented:
Third time's the charm? Not super convincing but there barely can be a significant regression, I feel, with the scrutiny exercised now...

```
# PR
tobias@td:~/go/src/github.com/cockroachdb/cockroach$ tail -n 2 artifacts/2f11e06d30/171003/kv95/enc\=false/nodes\=3/run_*/run_*kv.log | grep 600
  600.0s        0       22758429        37930.6      5.1      3.8     15.7     26.2    369.1  
  600.0s        0       22708048        37846.7      5.1      3.7     15.7     26.2     96.5  
  600.0s        0       22635711        37726.1      5.1      3.8     16.3     26.2    159.4  
  600.0s        0       22715758        37859.5      5.1      3.7     15.7     26.2    419.4  
  600.0s        0       22703217        37838.6      5.1      3.8     15.7     26.2    872.4  
# master base
tobias@td:~/go/src/github.com/cockroachdb/cockroach$ tail -n 2 artifacts/38fd4064f4/170952/kv95/enc\=false/nodes\=3/run_*/run_*kv.log | grep 600
  600.0s        0       22740384        37900.5      5.1      3.7     15.7     26.2    100.7  
  600.0s        0       22930005        38216.6      5.0      3.8     15.7     25.2    192.9  
  600.0s        0       22537494        37562.4      5.1      3.8     15.7     26.2    402.7  
  600.0s        0       22720525        37867.5      5.1      3.7     15.7     26.2    402.7  
  600.0s        0       21996586        36660.9      5.2      3.8     16.8     27.3    109.1  
```

tbg commented:
Also there's definitely something bad in CI. Repeated timeout in `make test` and

>  Execution timeout; tests failed: 15, passed: 58106, ignored: 246; failed to publish artifacts: Artifact file 'raw.0.json.txt' has size 19.06 GB which exceeds maximum allowed size of 3 GB. Maximum artifact size is configured at the Administration -> G... 

Hope it's some  dumb hang-on-shutdown thing, will find out.

tbg commented:
Yeah, looks like hang-on-shutdown. I can see tests stuck for 40+ minutes with stacks like these

```
goroutine 1658060 [select]:
github.com/cockroachdb/cockroach/pkg/util/retry.(*Retry).Next(0xc004942c30)
        /home/tobias/go/src/github.com/cockroachdb/cockroach/pkg/util/retry/retry.go:128 +0x14a
github.com/cockroachdb/cockroach/pkg/kv/kvclient/kvcoord.(*RangeIterator).Seek(0xc0049431f0, {0x59d06d0, 0xc004c9e870}, {0xc00b63a4b8, 0x0, 0x0},
 0x0)
        /home/tobias/go/src/github.com/cockroachdb/cockroach/pkg/kv/kvclient/kvcoord/range_iter.go:191 +0x44c
github.com/cockroachdb/cockroach/pkg/kv/kvclient/kvcoord.(*DistSender).divideAndSendBatchToRanges(0xc00a630480, {_, _}, {{{0x0, 0x0, 0x0}, 0x0, {
0x0, 0x0, 0x0, ...}, ...}, ...}, ...)
        /home/tobias/go/src/github.com/cockroachdb/cockroach/pkg/kv/kvclient/kvcoord/dist_sender.go:1164 +0x227
github.com/cockroachdb/cockroach/pkg/kv/kvclient/kvcoord.(*DistSender).Send(0xc00a630480, {0x59d06d0, 0xc004c9e870}, {{{0x0, 0x0, 0x0}, 0x0, {0x0
, 0x0, 0x0, ...}, ...}, ...})
        /home/tobias/go/src/github.com/cockroachdb/cockroach/pkg/kv/kvclient/kvcoord/dist_sender.go:804 +0x6d4
github.com/cockroachdb/cockroach/pkg/kv/kvclient/kvcoord.(*txnLockGatekeeper).SendLocked(0xc0077ace58, {0x59d06d0, 0xc004c9e870}, {{{0x0, 0x0, 0x
0}, 0x0, {0x0, 0x0, 0x0, ...}, ...}, ...})
```

```
goroutine 314295 [sleep]:
time.Sleep(0x4c4b40)
        /usr/local/go/src/runtime/time.go:193 +0x12e
github.com/cockroachdb/cockroach/pkg/util/stop.(*Stopper).Quiesce(0xc00a1de360, {0x59d06d0, 0xc000d2b830})
        /home/tobias/go/src/github.com/cockroachdb/cockroach/pkg/util/stop/stopper.go:609 +0x116
github.com/cockroachdb/cockroach/pkg/testutils/testcluster.(*TestCluster).stopServers.func1(0xc0053b6f70, 0xc00a1de360)
        /home/tobias/go/src/github.com/cockroachdb/cockroach/pkg/testutils/testcluster/testcluster.go:114 +0xda
created by github.com/cockroachdb/cockroach/pkg/testutils/testcluster.(*TestCluster).stopServers
        /home/tobias/go/src/github.com/cockroachdb/cockroach/pkg/testutils/testcluster/testcluster.go:110 +0x155
```

This is going to be somewhat annoying to fix (doable though) because any place with a retry loop basically needs access to the local stopper. We can't use error signaling here, since we do want to retry if the other node is shutting down; it's just when our local node is shutting down that we want to throw the towel. In other words, every caller of this will also want to check their stopper:

https://github.com/cockroachdb/cockroach/blob/d11a42c1f268fb31725438b678e1885f6643bd20/pkg/util/grpcutil/grpc_util.go#L111-L121

Even more so, really what we want is an inability to set up a retry loop without also checking the stopper before retries.

tbg commented:
A "right" solution here would be to have all contexts cancel on stopper drain. We're not anywhere near that, though I think we're at least intending to move into that direction (there is talk of replacing `context.Background` with `stopper.Background`, i.e. have all work in a running server derive from a stopper-owned context in some of @cockroachdb/cli's PRs). I could play whack-a-mole with all retry loops in the system to emulate it for now. I could also, instead, add gRPC client interceptors (i.e. hooks on outgoing RPCs) that check the local stopper and if quiescing, return an error that qualifies for `grpcutil.IsConnectionRejected`, which is our general purpose method we already call in such retry loops. I am gravitating towards the latter since it adds code in a central place and things won't rot the next time someone wants a new retry loop.

erikgrinaker commented:
>A "right" solution here would be to have all contexts cancel on stopper drain. We're not anywhere near that, though I think we're at least intending to move into that direction 

Yeah, this is definitely the way to go.

> I could play whack-a-mole with all retry loops in the system to emulate it for now.

Is this really going to be worth it though? Seems like it'd be about as much work overall as just propagating an appropriate context.

knz commented:
How would you feel about binding the AmbientContext to stopper, and have AnnotateCtx put a CancelOnQuiesce in there?

Asking in the context of my recent work on Ambientcontext

tbg commented:
> Is this really going to be worth it though? Seems like it'd be about as much work overall as just propagating an appropriate context.

Deriving the contexts from the stopper isn't so easy because a) the current way to cancel a context on stopper quiesce is to link the cancellation func into the stopper, and so there is a performance concern once we start doing this in places like `Replica.Send`. Also, we will likely need to plumb the stopper around a lot more in the first place. I'll leave that kind of thing to the server team, as I think it'll require more sustained effort than I can provide.

I will try something else, which is add an outgoing RPC intercept that fails fast with an error that we consider non-retriable (`IsConnectionRejected`) if our stopper is quiescing.

tbg committed:
rpc: reject outbound dial when quiescing

When a cluster (in particular a TestCluster) shuts down, the cluster
by definition will be unavailable. However, if we're not careful, we
may hang during shutdown since we have infinite retry loops in the
system that can only terminate when the cluster is available.

Prevent such situations (or at least make the default ones unlikely) by
failing outbound dial attempts *from a quiescing node* with an error
that our existing safeguards against infinitely stuck loops
(`IsAuthError` and `IsConnectionRejected`) interpret as non-retriable.

Release note: None

tbg commented:
> I will try something else, which is add an outgoing RPC intercept that fails fast with an error that we consider non-retriable (IsConnectionRejected) if our stopper is quiescing.

Did that but still seeing hangs. I have to investigate that tomorrow.
[test.log](https://github.com/cockroachdb/cockroach/files/7643041/test.log)

tbg commented:
One obvious source is here

https://github.com/cockroachdb/cockroach/blob/57ad801fb267fb7f2f594c28c6c3e958923d5ba0/pkg/ccl/kvccl/kvtenantccl/connector.go#L335-L357

I think here the tenant is not draining yet but the host cluster is, and in that case the tenant will see the "node unavailable" error which is not an AuthError.

I also see hangs outside of tenants so there's something else too, but this one reproduces 100% of the time with `make test PKG=./pkg/ccl/importccl/ TESTS=TestImportInTenant TESTFLAGS='-v -show-logs'` so I'll fix that first.

tbg committed:
kvserver: drop gossip subscriptions from tenants on quiesce

Tenants hold long-running GossipSubscription streaming RPCs. On the host
cluster side, these were not checking for (host cluster) stopper
cancellation. This is not an issue if these subscriptions aren't served
in a task but we're about to do so. Once we do that, there will be a
guaranteed deadlock in `stopper.Stop` as the task will not tear down.

Release note: None

tbg committed:
sidetransport: fix mockReceiver cancellation

Once we introduce stopper tasks wrapping the server-side RPCs,
the `sidetransport` test `TestRPCConnUnblocksOnStopper` will
hang without this commit, as `(*mockReceiver).PushUpdates` will be
invoked in a task, and thus the `Closer` that closes the blocking
channel would never be reached by the stopper.

Release note: None

tbg commented:
I think this will now, finally, turn green. RFAAHFL (Ready For Another And Hopefully Final Look 🤞🏽) @erikgrinaker.

erikgrinaker reviewed (approved):
> Reviewed 3 of 3 files at r8, 1 of 1 files at r9, 1 of 1 files at r10, 2 of 2 files at r11, 2 of 2 files at r12, 1 of 1 files at r13, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/73155)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @tbg)

___
*[pkg/rpc/context.go, line 1058 at r12](https://reviewable.io/reviews/cockroachdb/cockroach/73155#-MqPV1zOBLnVhIkEQHg6:-MqPV1zOBLnVhIkEQHg7:bkbby2n) ([raw file](https://github.com/cockroachdb/cockroach/blob/922234636b9b06f0576ad6824e37f2a6ca1d4c5b/pkg/rpc/context.go#L1058)):*
> ```Go
> 	conn, err := grpc.DialContext(rpcCtx.masterCtx, target, dialOpts...)
> 	if err != nil && rpcCtx.masterCtx.Err() != nil {
> 		// If node is draining, return special error (see its comments).
> ```

Consider spelling out that `RunAsyncTask` can only error if the stopper is quiescing. For a second I thought there was a missing conditional here or something.
___
*[pkg/rpc/context.go, line 84 at r13](https://reviewable.io/reviews/cockroachdb/cockroach/73155#-MqPW0BO9o5z7aby3dDO:-MqPW0BO9o5z7aby3dDP:b-bnp14x) ([raw file](https://github.com/cockroachdb/cockroach/blob/fa77fe86febeaffb642216d7d380e13c5f8ba3ca/pkg/rpc/context.go#L84)):*
> ```Go
> // `grpcutil.IsConnectionRejected` which prevents infinite retry loops during
> // cluster shutdown, especially in unit testing.
> var errDialRejected = grpcstatus.Error(codes.PermissionDenied, "refusing to dial; node is quiescing")
> ```

Maybe update the comments on `IsConnectionRejected` and `IsAuthError` to mention this case too.

It's a bit unfortunate that we're overloading these error codes that much, but I don't see any other codes that would be appropriate.


<!-- Sent from Reviewable.io -->

andreimatei reviewed (commented):
> A "right" solution here would be to have all contexts cancel on stopper drain.

FWIW, [this patch](https://github.com/cockroachdb/cockroach/pull/66387/commits/290765b77ba26c8652c87762e3df64fa74490ae5) was trying to at least add that as an option to the `RunAsyncTask` interface.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/73155#-:-MqPXzkY7K9BsG41Fz3c:b-n0r7dg)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @tbg)



<!-- Sent from Reviewable.io -->

tbg commented:
bors r=erikgrinaker
